### ISSUE TITLE
# TODO: why does this crash Julia?

### ISSUE THREAD

**Issue by mmiller-max** [JuliaSIMD/CheapThreads.jl#4]
Hey, loving the idea of this package and have been playing around with it today. Appreciate it's pretty new and in an early state, but it will be really useful for me.

I've been running into the same error that the commented-out test case has, and it seems to be something to do with the argument type. The only working types I've found are `Number`s and `Array{<:Number}`s. `String`s, `structs and `Array{Array{<:Number}}` cause Julia to crash. 

i.e. this works:
```Julia
function slow_task!((x, digits, n), j, k)
        start = 1 + (n * (j - 1)) ÷ num_threads()
        stop =  (n* k) ÷ num_threads()
        target = 0.0
        for i ∈ start:stop
            target += occursin(string(digits), string(i)) ? 0.0 : 1.0 / i
        end
        x[1,j] = target
    end
    function slow_cheap(n, digits)
        x = zeros(8, num_threads())
        batch(slow_task!, (num_threads(), num_threads()), x, digits, n)
        sum(@view(x[1,1:end]))
    end
    slow_cheap(1000, 9)
```
And this doesn't:
```
function slow_task!((x, digits, n), j, k)
        start = 1 + (n * (j - 1)) ÷ num_threads()
        stop =  (n* k) ÷ num_threads()
        target = 0.0
        for i ∈ start:stop
            target += occursin(digits, string(i)) ? 0.0 : 1.0 / i
        end
        x[1,j] = target
    end
    function slow_cheap(n, digits)
        x = zeros(8, num_threads())
        batch(slow_task!, (num_threads(), num_threads()), x, digits, n)
        sum(@view(x[1,1:end]))
    end
    slow_cheap(1000, "9")
```
Hope this helps debugging this issue, if you weren't already aware of this. Let me know if I can help with this at all!

**Comment by chriselrod** [comment]
Hi!
Did you by any chance see the slack discussion motivated by [this StackOverflow discussion](https://stackoverflow.com/questions/66343723/how-to-make-multi-threading-in-julia-scale-with-the-number-of-threads)?

I tried to get it to work with here, but didn't have the time to debug it when it didn't work.

I'd definitely like to get to the bottom of it though.
If you'd like to take a look/see if you can figure it out, a lot of the relevant code is in [StrideArraysCore/src/thread_compatible.jl](https://github.com/chriselrod/StrideArraysCore.jl/blob/master/src/thread_compatible.jl).

StrideArraysCore's [tests](https://github.com/chriselrod/StrideArraysCore.jl/blob/master/test/runtests.jl) include:
```julia
        xu = zeros(UInt, 100);
        x = rand(100); y = rand(100); z = rand(100);
        t = (x,y,z)
        pt, gt = StrideArraysCore.object_and_preserve(t)
        greet = ["Hello", "world", "hang", "in", "there"]
        pg, gg = StrideArraysCore.object_and_preserve(greet)
        ph, gh = StrideArraysCore.object_and_preserve(greet[1])
        pht, ght = StrideArraysCore.object_and_preserve((3,greet[1]))
        GC.@preserve xu gt gg gh ght begin
            ThreadingUtilities.store!(pointer(xu), pt, 0)
            @test @inferred(ThreadingUtilities.load(pointer(xu), typeof(pt), 0)) === (sizeof(UInt), t)
            offset = sizeof(UInt)
            for a ∈ t
                _p, g = StrideArraysCore.object_and_preserve(a)
                ThreadingUtilities.store!(pointer(xu), _p, offset)
                offset, p = @inferred(ThreadingUtilities.load(pointer(xu), typeof(_p), offset))
                @test g === a
                @test p isa PtrArray
                @test pointer(p) === pointer(a)
                @test p !== a
                @test p == a
            end
            ThreadingUtilities.store!(pointer(xu), pg, offset)
            @test @inferred(ThreadingUtilities.load(pointer(xu), typeof(pg), offset)) === (offset+sizeof(UInt), greet)
            ThreadingUtilities.store!(pointer(xu), ph, offset)
            @test @inferred(ThreadingUtilities.load(pointer(xu), typeof(ph), offset)) === (offset+sizeof(UInt), greet[1])
            ThreadingUtilities.store!(pointer(xu), pht, offset)
            @test @inferred(ThreadingUtilities.load(pointer(xu), typeof(pht), offset)) === (offset+sizeof(UInt), (3,greet[1]))
        end
```

Between `greet` and `greet[1]`, both a vector of strings and a string seem like they should work here.
But instead they crash Julia.


`CheapThreads.jl` works by sending jobs to tasks in the form of a function pointer, and storing all the needed data in a pointer.
The task then loads the function pointer and calls it.
That function then loads all the needed data from the pointer, and calls the actual function you want to run, passing the data as arguments.

`isbits` types work because you can just store them directly.
For `AbstractArray`s of `isbits`, `StrideArraysCore` tries to turn them into a `PtrArray`, which is just a pointer to the original data plus some relevant additional info like the array's size. It is built on [ArrayInterface.jl](https://github.com/SciML/ArrayInterface.jl), so it is fairly flexible in the types of arrays it supports (e.g. `Adjoint`, `PermutedDimsArray`s, `SubArray`s are supported, but special structures like `Diagonal`, `LowerTriangular`, etc aren't yet).

I thought I figured out how to safely load/store heap objects to get them to work in the StrideArraysCore tests.
But apparently there's an issue somewhere / something is going wrong.

A word of warning if you try debugging: IO locks up `ThreadingUtilities`/`CheapThreads`, so you can't print.
Something you can do though is
```julia
const OBJDUMP = Vector{Any}(undef, Threads.nthreads())
```
and then have your code write into `OBJDUMP[Threads.threadid()]` if you want to look at variables.
If you get an error instead of a crash, you should be able to use
```julia
ThreadingUtilites.reinitialize_tasks!()
```
to restart workers.

Beware that sometimes tasks live on after Julia crashes.
Worth checking `(h)top`/your task manager and killing still-running Julia tasks.

Please feel free to ask any questions!

**Comment by mmiller-max** [comment]
> Did you by any chance see the slack discussion

Unfortunately not, has it been lost to the slack hole? Couldn't find anything on a quick search.

I'll definitely try and take a look at this, thanks for the various pointers and I'll message with any updates/questions. FYI it's my first time looking at this sort of thing in detail so likely won't be super quick.. 

> A word of warning if you try debugging: IO locks up ThreadingUtilities/CheapThreads, so you can't print.

Ah yes I came across this, thanks for the debugging tip!

**Comment by chriselrod** [comment]
> Unfortunately not, has it been lost to the slack hole? Couldn't find anything on a quick search.

I just checked. The oldest thread I've participated in is merely 5 days old. It's gone.

> I'll definitely try and take a look at this, thanks for the various pointers and I'll message with any updates/questions. FYI it's my first time looking at this sort of thing in detail so likely won't be super quick..

Sounds good. Diving in tends to be a great way to learn. I'm happy to answer questions, and open to other suggestions if you (or anyone else) happens to have them.

**Comment by mmiller-max** [comment]
It seems to be this line where Julia crashes:

https://github.com/JuliaSIMD/CheapThreads.jl/blob/867bbfc26ff6654f90dd3313f9ec8d8dc80a3a66/src/batch.jl#L20

If I comment this out, or split it out into its elements like below it works:

```
@inline function setup_batch!(p::Ptr{UInt}, fptr::Ptr{Cvoid}, argtup, start::UInt, stop::UInt)
    offset = ThreadingUtilities.store!(p, fptr, sizeof(UInt))
    offset = ThreadingUtilities.store!(p, argtup[1], offset)
    offset = ThreadingUtilities.store!(p, argtup[2], offset)
    offset = ThreadingUtilities.store!(p, argtup[3], offset)
    offset = ThreadingUtilities.store!(p, start, offset)
    offset = ThreadingUtilities.store!(p, stop, offset)
    nothing
end
```
However if I save off `argtup` using the `OBJDUMP` method you suggested I can `store` and `load` it fine it fine (as expected from the `StrideArraysCore` tests):

```
julia> tid = 0x00000002
0x00000002

julia> p = ThreadingUtilities.taskpointer(tid)
Ptr{UInt64} @0x00007fcec3ad4040

julia> offset = ThreadingUtilities.store!(p, a, sizeof(UInt))
49

julia> (offset, args) = ThreadingUtilities.load(p, typeof(argtup), sizeof(UInt))
(49, ([0.0 5.0e-324 … 5.0e-324 0.0; 5.0e-324 0.0 … 5.0e-324 0.0; … ; 0.0 5.0e-324 … 5.0e-324 5.0e-324; 0.0 5.0e-324 … 5.0e-324 0.0], StrideArraysCore.Reference{String}("9"), 1000))

julia> args == a
true
```

So I am wondering if it is either something to do with:
- garbage collection (seems unlikely as all the args are preserved, and preserving `argtup` as well didn't make a difference), or 
- Somehow when the batch closure is created the type that is saved in `A` isn't stored correctly somehow

**Comment by mmiller-max** [comment]
> However if I save off argtup using the OBJDUMP method you suggested I can store and load it fine it fine (as expected from the StrideArraysCore tests):

Actually I think this is the problem, shouldn't the `string` be dereferenced here?

**Comment by mmiller-max** [comment]
OK whilst this doesn't explain the crash behaviour fully, here are two different things that fix the crash.

### 1. Modifying `ThreadingUtilities` so that it always iterates through a `Tuple`

This is definitely hacky and will likely break other things, but by forcing `store!` and `load` to iterate through a tuple (by commenting out [this line](https://github.com/JuliaSIMD/ThreadingUtilities.jl/blob/445f957e7fbaab582dcd65a759b8dd7e4c9996b1/src/utils.jl#L84) and modifying [this line](https://github.com/JuliaSIMD/ThreadingUtilities.jl/blob/445f957e7fbaab582dcd65a759b8dd7e4c9996b1/src/utils.jl#L95) to always return `:(_store!(p, tup, i))`), these functions use the extended methods in `StrideArraysCore` and Julia doesn't crash.

### 2. Calling `object_and_preserve` on all of `argtup`

As per the changes on my [branch](https://github.com/mmiller-max/CheapThreads.jl/tree/mm/crash-investigation), modifying `_batch_no_reserve` and `add_var` so that `object_and_preserve` is called on the whole of `argtup` seems to work. For the crashing unit test, the changes result in this code being generated:

```
argtup_temp = (args[1], args[2], args[3])
(argtup, gcargtup) = object_and_preserve(argtup_temp)
```

And then `argtup` is used as normal (well nearly, the dereferencing changes for the call to `f!` on thread 1). Doing this also uses the `store!` and `load` methods extended by `StrideArraysCore`. This also lines up more closely with [this test](https://github.com/JuliaSIMD/StrideArraysCore.jl/blob/fcb1c559b881f5553fa50c8289fc12ecc4e00d62/test/runtests.jl#L127-L128) in `StrideArraysCore`.

I think this approach is probably a good option for fixing it, but would be grateful to hear your thoughts on it. I can see that `args` are handled individually on purpose and guessing there is a good reason for this!

**Comment by chriselrod** [comment]
> I think this approach is probably a good option for fixing it, but would be grateful to hear your thoughts on it. I can see that args are handled individually on purpose and guessing there is a good reason for this!

Yes, and that's why I prefer option 1.
I'm not really sure what the memory layout of a nested tuple/of these non-isbits types.
Because of this:
```julia
julia> t = ("hi", "world", 9)
("hi", "world", 9)

julia> xu = zeroes(UInt,100);
ERROR: UndefVarError: zeroes not defined
Stacktrace:
 [1] top-level scope
   @ REPL[18]:1

julia> xu = zeros(UInt,100);

julia> Base.unsafe_store!(Base.unsafe_convert(Ptr{typeof(t)}, xu), t)
Ptr{Tuple{String, String, Int64}} @0x00007f6be5bb8050

julia> xu'
1×100 adjoint(::Vector{UInt64}) with eltype UInt64:
 0x00007f6bdf0cc510  0x00007f6bdf0cc530  0x0000000000000009  0x0000000000000000  0x0000000000000000  …  0x0000000000000000  0x0000000000000000  0x0000000000000000  0x0000000000000000  0x0000000000000000
```
I had thought that it was basically like storing pointers to each object, although this doesn't do what I'd want/expect in that case:
```julia
julia> Base.unsafe_load(Base.unsafe_convert(Ptr{typeof(t[1])}, xu))
```
(i.e., doesn't load the first string).

I'd thought that because of this, the tuple itself as a container doesn't matter. That all that mattered was the individual components, and therefore preserving them is fine.

Instead, it looks like I have to do that manually. I.e., option 1, store each element of the tuple individually, and then load them and reconstruct the tuple.

Why do I prefer that over option 2?
Two reasons:

1. Because I want the ability to handle some types specially. E.g., certain `AbstractArray`s -> `PtrArray`s. I guess this isn't working at the moment, but would if we do option `1.`
2. Because `GC.@preserve`-ing the combined tuple often allocates memory. If we suppose that it is an object that is supposed to "actually exist" in memory, and hence disappearing causes crashes, it is also understandable why `GC.@preserve`-ing it would force heap allocation, thereby causing allocations (although mutable structs only holding `isbits` objects do not allocate when you `GC.@preserve` them, they're happy to live on the stack, only getting heap allocated if they escape). Either way, I do empirically observe allocations from `GC.@preserve`-ing structs holding `!isbits`, and therefore have been taking the approach of deconstructing and reconstructing them to avoid these allocations whenever possible. 

tldr: I'm not exactly sure either, but option 1 is empirically preferable from a performance-perspective and we can speculate on why that would make sense.

> I'm still not exactly sure what is causing the crash, as store! and load can be used fine with argtup in the REPL, but for some reason they break whilst trying to do the threading.

Something to try would be doing it in a function without threading. I think it's likely we'll get crashes then.
I.e., that the problem isn't theading, but being in a function that lets `argtup` get optimized, and forcing us to `GC.@preserve` it. If `argtup` is live in the REPL, it's safe from the GC and doesn't need preserving.


> This is definitely hacky and will likely break other things, but by forcing `store!` and `load` to iterate through a tuple (by commenting out [this line](https://github.com/JuliaSIMD/ThreadingUtilities.jl/blob/445f957e7fbaab582dcd65a759b8dd7e4c9996b1/src/utils.jl#L84) and modifying [this line](https://github.com/JuliaSIMD/ThreadingUtilities.jl/blob/445f957e7fbaab582dcd65a759b8dd7e4c9996b1/src/utils.jl#L95) to always return `:(_store!(p, tup, i))`), these functions use the extended methods in `StrideArraysCore` and Julia doesn't crash.

I think that's fine. Mind making a PR?

**Comment by mmiller-max** [comment]
Great, that makes sense. Thanks for your detailed explanation! I did see an allocation I didn't expect when doing a test just now too.

> Something to try would be doing it in a function without threading. I think it's likely we'll get crashes then.
I.e., that the problem isn't theading, but being in a function that lets `argtup` get optimized, and forcing us to `GC.@preserve` it. If `argtup` is live in the REPL, it's safe from the GC and doesn't need preserving.

I'll give this a go, interesting to test.

> I think that's fine. Mind making a PR?

Will get one ready

**Comment by mmiller-max** [comment]
Is this the sort of test you had in mind?

```
using ThreadingUtilities

function f_test(argtup)
    p = f_store(argtup)
    A = typeof(argtup)
    f_load(p, A)
end

function f_store(argtup)
    new_args = deepcopy(argtup) # Hopefully stops it being preserved
    tid = 0x00000002
    p = ThreadingUtilities.taskpointer(tid)
    offset = ThreadingUtilities.store!(p, new_args, sizeof(UInt))
    return p
end

function f_load(p, A)
    (offset, args) = ThreadingUtilities.load(p, A, sizeof(UInt))
    return args
end

args = f_test(argtup)
``
This also doesn't crash I'm doing it like this as I'm pulling `argtup` out of `OBJDUMP`, which means it is in the REPL. I hoped the deepcopy would result in it being optimised away when the function ended but it didn't seem to.

**Comment by chriselrod** [comment]
Almost, just two things missing:
1. The deepcopy guarantees that it is temporarily "preserved". Look at the `@code_typed` and `@code_llvm` and you'll see that `deepcopy_internal` is not inlined. Being passed to a function that isn't inlined forces it to be allocated. Because `new_args` escaped the non-inlined function, it too must've been allocated.
2. No need to preserve a variable that exists in the global scope; `argtup` as implied in your example already exists as a global variable. It won't get optimized away. 

So we (a) remove the `deepcopy` and (b)  create `argtup` inside a function:
```julia
using ThreadingUtilities

function f_test(argtup)
    p = f_store(argtup)
    A = typeof(argtup)
    f_load(p, A)
end

function f_store(argtup)
    tid = 0x00000002
    p = ThreadingUtilities.taskpointer(tid)
    offset = ThreadingUtilities.store!(p, argtup, sizeof(UInt))
    return p
end

function f_load(p, A)
    (offset, args) = ThreadingUtilities.load(p, A, sizeof(UInt))
    return args
end

f_test(("hi", "world"))
f_test_2() = f_test(("hi", "world"))
f_test_2()
```
I get:
```julia
julia> f_test(("hi", "world"))
("hi", "world")

julia> f_test_2() = f_test(("hi", "world"))
f_test_2 (generic function with 1 method)

julia> f_test_2()
julia: /home/chriselrod/Documents/languages/julia/src/cgutils.cpp:1520: llvm::Value* data_pointer(jl_codectx_t&, const jl_cgval_t&): Assertion `x.ispointer()' failed.

signal (6): Aborted
in expression starting at REPL[7]:1
```

**Comment by mmiller-max** [comment]
That did it, thanks for explaining! One last question - do we know that `deepcopy_internal` is not inlined because of the `invoke` in `@code_typed`?

**Comment by chriselrod** [comment]
> That did it, thanks for explaining! One last question - do we know that `deepcopy_internal` is not inlined because of the `invoke` in `@code_typed`?

`@code_typed` shows us code after Julia runs its optimizations. Julia handles all inlining for Julia code (it runs LLVM's "alwaysinline" pass, but not the normal inlining pass), so if it hasn't inlined into `@code_typed`, there is no later stage in the pipeline that will inline it.

**Comment by mmiller-max** [comment]
OK makes sense, thanks

**Comment by mmiller-max** [comment]
Do you want a PR to increment the compat for `ThreadingUtilities` or are you happy just to close this off for now?

**Comment by chriselrod** [comment]
> Do you want a PR to increment the compat for `ThreadingUtilities` or are you happy just to close this off for now?

Sure.
